File: DSA/BST/python/bst.py
#!/usr/bin/env python3
"""
    Implementation of Binary Search Tree Data Structure.
"""

class Node:
    """Representation of a Node in the Binary Search Tree."""
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BinarySearchTree:
    """Representation of Binary Search Tree."""
    def __init__(self):
        self.root = None

    def insert(self, value):
        """
        Insert a new node with the given value into the tree.
        
        Args:
            value: Value of the new node to be inserted.
        Returns:
            The BinarySearchTree instance.
        """
        new_node = Node(value)

        if self.root is None:
            self.root = new_node
            return self

        current = self.root
        while True:
            if value == current.value:
                return None
            if value < current.value:
                if current.left is None:
                    current.left = new_node # type: ignore
                    return self
                current = current.left
            else:
                if current.right is None:
                    current.right = new_node # type: ignore
                    return self
                current = current.right

    def contains(self, value):
        """
        Check if a value exists in the tree.
        
        Args:
            value: Value to check.
        Returns:
            Boolean indicating if the value is found.
        """
        if self.root is None:
            return False

        current = self.root
        while current:
            if value < current.value:
                current = current.left
            elif value > current.value:
                current = current.right
            else:
                return True
        return False

File: DSA/BST/javascript/bst.js
#!/usr/bin/env node
/**
 * Implementation of Binary Search Tree Data Structure.
 */

/**
 * Representation of a Node into the Binary Search Tree.
 */
class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

/**
 * Representation of Tree (Binary Search Tree).
 */
class BinarySearchTree {
  constructor() {
    this.root = null;
  }

  /**
   *
   * @param {*} value : Value of the New Node to be inserted.
   * @returns : The whole Tree.
   */
  insert(value) {
    var newNode = new Node(value);

    if (this.root === null) {
      this.root = newNode;
      return this;
    }

    var current = this.root;
    while (true) {
      if (value === current.value) return undefined;
      if (value < current.value) {
        if (current.left === null) {
          current.left = newNode;
          return this;
        }
        current = current.left;
      } else {
        if (current.right === null) {
          current.right = newNode;
          return this;
        }
        current = current.right;
      }
    }
  }

  /**
   *
   * @param {*} value : Value to check if its in Tree or Not.
   * @returns : Boolean based on the value if Found or not.
   */
  contains(value) {
    if (this.root === null) return false;
    let current = this.root;
    let found = false;
    while (current && !found) {
      if (value < current.value) {
        current = current.left;
      } else if (value > current.value) {
        current = current.right;
      } else {
        return true;
      }
    }
    return false;
  }
}
